xilem/examples/calc.rs

```diff
@@ -0,0 +1,326 @@
+    }
+
+    fn on_equals(&mut self) {
+        // Requires both numbers be present
+        if self.numbers[0].is_empty() || self.numbers[1].is_empty() {
+            play_bell_sound();
+            return; // Just abort.
+        } else if self.result.is_some() {
+            // Repeat the operation using the prior result on the left.
+            self.numbers[0] = self.result.clone().unwrap();
+        }
+        self.current_num_index = 0;
+        let num1 = self.numbers[0].parse::<f64>();
+        let num2 = self.numbers[1].parse::<f64>();
+        // Display format error or display the result of the operation.
+        self.result = Some(match (num1, num2) {
+            (Ok(num1), Ok(num2)) => self.operation.unwrap().perform_op(num1, num2).to_string(),
+            (Err(err), _) => err.to_string(),
+            (_, Err(err)) => err.to_string(),
+        });
+    }
+
+    fn on_delete(&mut self) {
+        if self.result.is_some() {
+            // Delete does not do anything with the result.
+            play_bell_sound();
+            return;
+        }
+        let mut num = self.get_current_number();
+        if !num.is_empty() {
+            num.remove(num.len() - 1);
+            self.set_current_number(num);
+        } else {
+            play_bell_sound();
+        }
+    }
+
+    fn negate(&mut self) {
+        // If there is a result, negate that after clearing and moving it to the first number
+        if self.result.is_some() {
+            self.move_result_to_left();
+        }
+        let mut num = self.get_current_number();
+        if num.is_empty() {
+            play_bell_sound();
+            return;
+        }
+        if num.starts_with('-') {
+            num.remove(0);
+        } else {
+            num = format!("-{num}");
+        }
+        self.set_current_number(num);
+    }
+}
+
+const DISPLAY_FONT_SIZE: f32 = 30.;
+const GRID_GAP: f64 = 2.;
+fn app_logic(data: &mut AppData) -> impl WidgetView<AppData> {
+    flex((
+        // Display
+        flex((
+            FlexSpacer::Flex(0.1),
+            label(data.numbers[0].clone()).text_size(DISPLAY_FONT_SIZE),
+            data.operation
+                .map(|operation| label(operation.as_str()).text_size(DISPLAY_FONT_SIZE)),
+            label(data.numbers[1].clone()).text_size(DISPLAY_FONT_SIZE),
+            data.result
+                .clone()
+                .map(|_| label("=").text_size(DISPLAY_FONT_SIZE)),
+            data.result
+                .clone()
+                .map(|result| label(result).text_size(DISPLAY_FONT_SIZE)),
+            FlexSpacer::Flex(0.1),
+        ))
+        .direction(Axis::Horizontal)
+        .cross_axis_alignment(CrossAxisAlignment::Center)
+        .main_axis_alignment(MainAxisAlignment::Start)
+        .gap(5.)
+        .flex(1.0),
+        FlexSpacer::Fixed(10.0),
+        // Top row
+        flex((
+            sized_box(button("CE", |data: &mut AppData| data.clear_entry()))
```

Comment:
I agree, I think all those buttons could probably summed up as one function such as:

`fn calc_button(action: CalcAction) -> AnyFlexChild<AppData>`
or 
`fn calc_button(action: CalcAction) -> FlexItem<impl WidgetView<AppData>, AppData>`
to avoid type erasure.
Only calling one function like `AppData::execute(&mut self, CalcAction)`.

and `fn calc_row(row: [CalcAction; 3]) -> impl WidgetView<AppData>` to reduce the boilerplate for rows, with something like

`flex(row.map(calc_button))` (plus rest of the params of the `Flex` view)
as content.

This would lead to a little bit more elm like code, slightly more data-driven, I guess.
Not sure whether that's good or bad.
But it could also serve as an example for creating some logging/recording actions possibly interesting for something like undo/redo, with `CalcAction` being an enum:

```rust
enum CalcAction {
    AddDigit(u8),
    AddSeparator,
    BinaryOperation(MathOperator),
    ...
}
```
(To support undo/redo, either the whole log is evaluated at once each time an action happens, and there's an array-index pointing to the current undo-snapshot, or there are variants that can be converted interchangeably like `RemoveDigit(u8)` <-> `AddDigit(u8)` with a `fn undo_action(CalcAction) -> CalcAction`)

Btw. not that this is an actual suggestion, more a tip to reduce this a little bit:
```suggestion
            sized_box(button("CE", AppData::clear_entry))
```

Reply:
Hmm. I think that the `CalcAction` suggestion might be over-abstracting.

Reply:
Maybe, but also maybe closer to a real app (with something like a command based undo/redo system, which I find interesting to think about at least, and it certainly helps reducing the ui boilerplate for the buttons to two short functions).

I guess this could also done slightly different, alternatively, all the buttons that don't take additional parameters, could be something like:

```rust
fn command_button(label, impl Fn(&mut AppData)) -> FlexItem {...}
fn operator_button(MathOperator) -> FlexItem {...}
fn digit_button(digit) -> FlexItem {...}

// Could also accept an impl ViewSequence<Element = FlexElement> I guess,
// but not sure if we wan't to "expose" `ViewSequence` here...,
// we could also create a `FlexSequence` similar as `DomView` is to `View`, but the marker is annoying for something like that, as it has to be provided as generic parameter...
fn button_row(buttons: (impl FlexItem, impl FlexItem, impl FlexItem, impl FlexItem)) -> impl WidgetView
```

or to avoid all the generic parameters, it could also be `AnyFlexChild` for each of those parameters and the button functions returning `AnyFlexChild`.
so

```rust
fn button_row(buttons: [AnyFlexChild<AppData>; 4]) -> impl WidgetView
```

Then something like
`num_button_row(...)` to reduce/reuse further (if really useful).